filebeat/input/journald/pkg/journalctl/journalctl.go

```diff
@@ -0,0 +1,146 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package journalctl
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+
+	input "github.com/elastic/beats/v7/filebeat/input/v2"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type journalctl struct {
+	cmd      *exec.Cmd
+	dataChan chan []byte
+	errChan  chan string
+	stdout   io.ReadCloser
+	stderr   io.ReadCloser
+
+	logger   *logp.Logger
+	canceler input.Canceler
+}
+
+// Factory returns an instance of journalctl ready to use.
+//
+// The returned type is an interface to allow mocking for testing
+func Factory(canceller input.Canceler, logger *logp.Logger, binary string, args ...string) (Jctl, error) {
+	cmd := exec.Command(binary, args...)
+
+	jctl := journalctl{
+		canceler: canceller,
+		cmd:      cmd,
+		dataChan: make(chan []byte),
+		errChan:  make(chan string),
+		logger:   logger,
+	}
+
+	var err error
+	jctl.stdout, err = cmd.StdoutPipe()
+	if err != nil {
+		return &journalctl{}, fmt.Errorf("cannot get stdout pipe: %w", err)
+	}
+	jctl.stderr, err = cmd.StderrPipe()
+	if err != nil {
+		return &journalctl{}, fmt.Errorf("cannot get stderr pipe: %w", err)
+	}
+
+	go func() {
+		defer jctl.logger.Debug("stderr reader goroutine done")
+		defer close(jctl.errChan)
+		reader := bufio.NewReader(jctl.stderr)
+		for {
+			line, err := reader.ReadString('\n')
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					return
+				}
+				logger.Errorf("cannot read from journalctl stderr: %s", err)
+				return
+			}
+
+			logger.Errorf("Journalctl wrote to stderr: %s", line)
+		}
+	}()
+
+	// Goroutine to read events from stdout
+	go func() {
+		defer jctl.logger.Debug("stdout reader goroutine done")
+		defer close(jctl.dataChan)
+		reader := bufio.NewReader(jctl.stdout)
+		for {
+			data, err := reader.ReadBytes('\n')
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					return
+				}
+				logger.Errorf("cannot read from journalctl stdout: %s", err)
+				return
+			}
+
+			select {
+			case <-jctl.canceler.Done():
+				return
+			case jctl.dataChan <- data:
+			}
+		}
+	}()
+
+	logger.Infof("Journalctl command: journalctl %s", strings.Join(args, " "))
+
+	if err := cmd.Start(); err != nil {
+		return &journalctl{}, fmt.Errorf("cannot start journalctl: %w", err)
+	}
+
+	logger.Infof("journalctl started with PID %d", cmd.Process.Pid)
+
+	go func() {
+		if err := cmd.Wait(); err != nil {
```

Comment:
I'm wondering if we should also watch for the `dataChan` and `errChan` to close.   For example, if we failed to read stdout for some reason, that go routine loop would exit

Reply:
This goroutine is here only to log any error returned by `cmd.Wait`, if the process exits unexpectedly its stderr and stdout will be closed and the reader gorourines will get an EOF or error. At the moment I'm doing the best to read and ship/log all data without over complicating the code and risking getting into a dead lock.